Buiding a movies recommendation system using
scikit-suprise.

ML algorithm == k-Nearest Neighbours.

Dataset == ml-100k built in suprise module.

In [36]:
# import the necessary tools
import pandas as pd
from surprise import Dataset
from surprise import KNNWithMeans                         



In [37]:
# load the dataset,
data = Dataset.load_builtin('ml-100k')                     


In [38]:
# view the dataset as a pandas dataframe.                                  
frame = pd.DataFrame(data.__dict__['raw_ratings'], columns=['user_id','item_id','rating','timestamp'])                   
frame.head()


user_id item_id  rating  timestamp
0     196     242     3.0  881250949
1     186     302     3.0  891717742
2      22     377     1.0  878887116
3     244      51     2.0  880606923
4     166     346     1.0  886397596

In [39]:
# more broad view.        

frame

user_id item_id  rating  timestamp
0         196     242     3.0  881250949
1         186     302     3.0  891717742
2          22     377     1.0  878887116
3         244      51     2.0  880606923
4         166     346     1.0  886397596
...       ...     ...     ...        ...
99995     880     476     3.0  880175444
99996     716     204     5.0  879795543
99997     276    1090     1.0  874795795
99998      13     225     2.0  882399156
99999      12     203     3.0  879959583

[100000 rows x 4 columns]

In [40]:
# building the the ml algorithm.                                                    
params = {
  "name": "cosine",
  "user_based": False # compute similarity between items and not users.
}

algo = KNNWithMeans(sim_options=params)



In [41]:
# Seperate the trainset from the data.                                                               
trainset = data.build_full_trainset()

trainset                   


In [42]:
# train the ml algo.                                                         

algo.fit(trainset)

Computing the cosine similarity matrix...
Done computing similarity matrix.


What is left is to make some predictions.

What the trained algo will be predicting is actually the ratings a user is likely to give to a new movie he/she hasn't watch yet.

Applications can now use the predicted ratings to recommend the movie to the user.

High rating signifying the movie will be recommended and vice-Versa.


we can play with the ml algo, by selecting a random rating, predict what the user would rate the movie and then compare it to the original rate.

In [43]:
# select a random rating
ratings = frame.iloc[100]
ratings


user_id             32
item_id            294
rating             3.0
timestamp    883709863
Name: 100, dtype: object

In [44]:
# make a prediction.  

user = ratings.user_id
item = ratings.item_id
rate = ratings.rating
timestamp = ratings.timestamp

pred = algo.predict(user, item, timestamp)
pred.est

2.9502425744709067

hurray!!
The prediction is closed.

rating == 3.0
predicted rating == 2.95

we did it.
But this is not enough assurance though, there is still need for performance scoring, params-tuning and validation test.

